src/server/editorServices.ts

```diff
@@ -1037,6 +1053,97 @@ namespace ts.server {
+                : globalCachePath;
+
+            // For inferred project, we always enable the auto typing discover feature, therefore the typingOptions
+            // has a default value
+            const typingOptions: TypingOptions =
+                project.projectOptions ? project.projectOptions.typingOptions : { enableAutoDiscovery: true };
+            const compilerOptions = project.projectOptions ? project.projectOptions.compilerOptions : undefined;
+
+            const { cachedTypingPaths, newTypingNames } = ts.JsTyping.discoverTypings(
+                sys,
+                project.getFileNames(),
+                globalCachePath,
+                cachePath,
+                typingOptions,
+                compilerOptions
+            );
+
+            // Bail out when the autoTyping is disabled
+            if (cachedTypingPaths.length === 0 && newTypingNames.length === 0) {
+                return;
+            }
+
+            if (!sys.directoryExists(cachePath)) {
+                sys.createDirectory(cachePath);
+            }
+            const tsdJsonPath = ts.combinePaths(cachePath, "tsd.json");
+            const typingsPath = ts.combinePaths(cachePath, "typings");
+            const api = tsd.getAPI(tsdJsonPath);
+
+            const options = new tsd.Options();
+            options.resolveDependencies = true;
+            options.overwriteFiles = true;
+            options.saveToConfig = true;
+
+            this.log("Cached typings include: " + cachedTypingPaths);
+            cachedTypingPaths.forEach(p => addTypingToProject(p, project));
+            project.projectService.updateProjectStructure();
+
+            this.log("New typings to download: " + newTypingNames);
+            if (newTypingNames && newTypingNames.length > 0) {
+                const query = new tsd.Query();
+                for (const newTypingName of newTypingNames) {
+                    query.addNamePattern(newTypingName);
+                }
+
+                let promise: PromiseLike<void>;
+                if (!sys.fileExists(tsdJsonPath)) {
+                    promise = api.initConfig(/*overwrite*/true).then((paths: string[]) => api.readConfig());
+                }
+                else {
+                    promise = api.readConfig();
+                }
+
+                promise
+                    .then(() => api.select(query, options))
+                    .then((selection: any) => api.install(selection, options))
+                    .then((installResult: any) => {
+                        const keys = Object.keys(installResult.written.dict);
+                        keys.forEach(key => addTypingToProject(ts.combinePaths(typingsPath, key), project));
+                        project.projectService.updateProjectStructure();
+                        if (cachePath !== globalCachePath) {
+                            JsTyping.updateNotFoundTypingNames(newTypingNames, cachePath);

```

Comment:
Why don't we also track not found names in the global cache to avoid retrying repeatedly there? (Add a comment to explain, as it's non-obvious).